src/Revise.jl

```diff
@@ -94,6 +94,74 @@ This list gets populated by callbacks that watch directories for updates.
 """
 const revision_queue = Set{Tuple{PkgData,String}}()
 
+const revision_event = Condition()
+
+"""
+    Revise.user_callbacks_queue
+
+Global variable, `user_callbacks_queue` holds `key` values for which the
+file has changed but the user hooks have not yet been called.
+"""
+const user_callbacks_queue = Set{Any}()
+
+const user_callbacks_by_file = Dict{String, Set{Any}}()
+const user_callbacks_by_key = Dict{Any, Any}()
+
+function add_callback(f, files, modules=nothing; key=gensym())
+    remove_callback(key)
+
+    files = map(abspath, files)
+    init_watching(files)
+
+    if modules !== nothing
+        for mod in modules
+            id = PkgId(mod)
+            pkgdata = pkgdatas[id]
+            for file in srcfiles(pkgdata)
+                absname = joinpath(basedir(pkgdata), file)
+                push!(files, absname)
+                track(mod, absname)
+            end
+        end
+    end
+
+    for file in files
+        cb = get!(Set, user_callbacks_by_file, file)
+        push!(cb, key)
+        user_callbacks_by_key[key] = f
+    end
+
+    return key
+end
+
+function remove_callback(key)
+    for cbs in values(user_callbacks_by_file)
+        delete!(cbs, key)
+    end
+    delete!(user_callbacks_by_key, key)
+    # TODO: stop watching these files
```

Comment:
Worth contemplating what happens if two callbacks get registered on an overlapping set of files.

Reply:
I just pushed a commit that just leaves the watches hanging around (even if they won't trigger callbacks any more). I added my thinking as a comment to the code:
```
# possible future work: we may stop watching (some of) these files
# now. But we don't really keep track of what background tasks are running
# and Julia doesn't have an ergonomic way of task cancellation yet (see
# e.g.
#     https://github.com/JuliaLang/Juleps/blob/master/StructuredConcurrency.md
# so we'll omit this for now. The downside is that in pathological cases,
# this may exhaust inotify resources.
```
What do you think?

Reply:
For any task that needs rescheduling it could of course just fail to reschedule, but we would have to contend with one more execution. I'm fine with leaving it this way, at least until people complain (this is not functionality I tend to use a lot myself).